src/components/input/test/Select-test.js

```diff
@@ -425,12 +432,83 @@ describe('Select', () => {
     });
   });
 
+  context('when listbox does not fit in available space', () => {
+    async function getOpenListbox(wrapper) {
+      toggleListbox(wrapper);
+      // Wait for listbox to be open
+      await waitFor(() => !isListboxClosed(wrapper));
+
+      return getListbox(wrapper);
+    }
+
+    it('never renders a listbox bigger than the viewport', async () => {
+      const name = 'name'.repeat(1000);
+      const wrapper = createComponent(
+        { buttonContent: 'Select a value' },
+        {
+          items: ['1', '2', '3'].map(id => ({ id, name })),
+        },
+      );
+
+      const listbox = await getOpenListbox(wrapper);
+      const { width: listboxWidth } = listbox
+        .getDOMNode()
+        .getBoundingClientRect();
+
+      assert.isTrue(listboxWidth < window.window.innerWidth);
+    });
+
+    [
+      { name: 'long name'.repeat(50), shouldGrow: true },
+      { name: 'short name', shouldGrow: false },
+    ].forEach(({ name, shouldGrow }) => {
+      it('makes listbox grow to the left if there is space', async () => {
+        const wrapper = createComponent(
+          { buttonContent: 'Select a value' },
+          {
+            items: ['1', '2', '3'].map(id => ({ id, name })),
+          },
+        );
+
+        const listbox = await getOpenListbox(wrapper);
+        const listboxStyleLeft = listbox.getDOMNode().style.left;
+        const listboxLeft = Number(listboxStyleLeft.replace('px', ''));
+        const { left: buttonLeft } = getToggleButton(wrapper)
+          .getDOMNode()
+          .getBoundingClientRect();
+
+        assert.equal(listboxLeft < buttonLeft, shouldGrow);
+      });
+    });
+    [
+      { name: 'long name'.repeat(500), shouldGrow: true },
+      { name: 'short name', shouldGrow: false },
+    ].forEach(({ name, shouldGrow }) => {
+      it('makes right Select listbox grow to the right if there is space', async () => {
+        const wrapper = createComponent(
+          { right: true, buttonContent: 'Select a value' },
+          {
+            items: ['1', '2', '3'].map(id => ({ id, name })),
+          },
+        );
+
+        const listbox = await getOpenListbox(wrapper);
+        const listboxStyleLeft = listbox.getDOMNode().style.left;
+        const listboxLeft = Number(listboxStyleLeft.replace('px', ''));
+
+        // 8 is the minimum `left` value that acts as a gap with the edge of the
+        // viewport
+        assert.equal(listboxLeft === 8, shouldGrow);
```

Comment:
This might be better written as two separate asserts, called according to `shouldGrow`. One branch would test `assert.equal(listboxLeft, 8)` the other a not-equal version of that. This will result in a more helpful error if the test fails.

Reply:
I refactored this test to test for the exact left and right coordinates of the listbox in three different cases:

* Small content: the listbox matches the button size due to the `min-width` CSS prop.
* Medium content: the listbox is a bit bigger than the button and grows to the right.
* Big content: the listbox grows to the very right side (considering the gap) and then grows a bit further to the right than the toggle button.